core/src/MpiShmem/Kokkos_MpiShmem_Impl.cpp

```diff
@@ -0,0 +1,251 @@
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact  H. Carter Edwards (hcedwar@sandia.gov)
+// 
+// ************************************************************************
+//@HEADER
+*/
+
+/*--------------------------------------------------------------------------*/
+/* Kokkos interfaces */
+
+#include <Kokkos_Core.hpp>
+
+/* only compile this file if MPISHMEM is enabled for Kokkos */
+#ifdef KOKKOS_HAVE_MPISHMEM
+
+#include <impl/Kokkos_Error.hpp>
+
+/*--------------------------------------------------------------------------*/
+/* Standard 'C' libraries */
+#include <stdlib.h>
+
+/* Standard 'C++' libraries */
+#include <iostream>
+
+/*--------------------------------------------------------------------------*/
+
+namespace Kokkos {
+namespace Impl  {
+
+// put MpiShmemInternal here as a singleton
+class MpiShmemInternal {
+private:
+  MpiShmemInternal( const MpiShmemInternal & );
+  MpiShmemInternal & operator = ( const MpiShmemInternal & );
+public:
+
+  typedef MpiShmem::size_type size_type;
+
+  MPI_Comm m_world;
+  MPI_Comm m_team;
+  bool     m_called_mpi_init;
+
+  static MpiShmemInternal & singleton();
+
+  int verify_is_initialized( const char * const label ) const;
+
+  int is_initialized() const
+  {
+    return m_world != MPI_COMM_NULL && m_team != MPI_COMM_NULL;
+  }
+
+  void initialize( MPI_Comm world, MPI_Comm team );
+  void finalize();
+
+  void print_configuration( std::ostream & ) const ;
+
+  void fence() const;
+
+  ~MpiShmemInternal();
+
+  MpiShmemInternal()
+    : m_world( MPI_COMM_NULL )
+    , m_team( MPI_COMM_NULL )
+    , m_called_mpi_init( false )
+  {}
+};
+
+MpiShmemInternal & MpiShmemInternal::singleton()
+{
+  static MpiShmemInternal self;
+  return self;
+}
+
+void MpiShmemInternal::initialize( MPI_Comm world, MPI_Comm team )
+{
+  int is_mpi_initialized;
+  MPI_Initialized(&is_mpi_initialized);
+  if (!is_mpi_initialized) {
+    MPI_Init(NULL,NULL);
+    m_called_mpi_init = true;
+  }
+  m_world = world;
+  if (team == MPI_COMM_NULL)
+    MPI_Comm_split_type(world, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &team);
```

Comment:
For other readers: This splits the "world" communicator into subcommunicators, each of which contains processes that can create shared-memory regions together.  Thus, this call lets MPI do the work of identifying the process equivalent of "thread teams."